### ISSUE TITLE
[BUG] Too many parallel network connections open while installation

### ISSUE THREAD

**Issue by ankurparihar** [npm/cli#7272]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### This issue exists in the latest npm version

- [X] I am using the latest npm

### Current Behavior

Hi,

For a project npm install opens nearly ~2100 connections in parallel, the same project installed via pnpm only takes about 18 concurrent connection.

This causes a huge bandwidth issue for our internal CI/CD pipelines.

### Expected Behavior

Please provide a way to limit concurrent connection.

### Steps To Reproduce

1. In this environment
  - MacOS / Linux
2. First clean cache
  - npm cache clean --force
  - pnpm store prune
2. Run `npm i` / `pnpm i` in any project containing package.json
5. in parallel run `watch -n 0.2 "ss -tunap | wc -l"` - to monitor open connections

### Environment

- npm: `10.2.4`
- Node.js: `v20.11.1`
- OS Name: `MacOS Sonoma 14.2.1 (23C71)`
- System Model Name: `Macbook Pro`
- npm config:
```ini
; node bin location = /usr/local/bin/node
; node version = v20.11.1
; npm local prefix = /home
; npm version = 10.2.4
; cwd = /home
; HOME = /root
; Run `npm config ls -l` to show all defaults.
```

**Comment by ankurparihar** [comment]
Update:

- Node 16.20.2, NPM: 8.19.4 ~25 max concurrent connections
- NODE: 18.19.1, NPM: 10.2.4 also makes ~2100 connections

It seems this was increased from Node 18

**Comment by ankurparihar** [comment]
It seems maxsockets is also not being respected, setting this to a lower value like 10 or 1 didn't have any effect.

**Comment by mhdawson** [comment]
I'm not sure if this is related, but we've seen some teams running into problems moving between versions of Node.js with npm 9 versus 10 and running out of file handles which sounds like it could be related.

**Comment by lukekarrys** [comment]
> Update:
> 
> 
> 
> - Node 16.20.2, NPM: 8.19.4 ~25 max concurrent connections
> 
> - NODE: 18.19.1, NPM: 10.2.4 also makes ~2100 connections
> 
> 
> 
> It seems this was increased from Node 18
> 
> 

Thanks for the report, I will take a look at this.

**Comment by ChristophLindemann** [comment]
@lukekarrys check out #7072 which has been open since December. It might have some more insights.

**Comment by ankurparihar** [comment]
Thank you for the prompt response, let me know if I can do some help, even in testing and all.

**Comment by melroy89** [comment]
> > Update:
> > 
> > * Node 16.20.2, NPM: 8.19.4 ~25 max concurrent connections
> > * NODE: 18.19.1, NPM: 10.2.4 also makes ~2100 connections
> > 
> > It seems this was increased from Node 18
> 
> Thanks for the report, I will take a look at this.

That would be much appreciated, I'm using Node LTS (v20.11.1), which is using NPM v10.2.4. Where this problem is still present, and causing a lot of issues during CI/CD pipelines.

**Comment by silverwind** [comment]
Imho, it's definitely related to `maxsockets` having not effect in npm 10.

**Comment by dominicfraser** [comment]
@neelance commented something interesting in https://github.com/npm/cli/issues/7072

> [neelance](https://github.com/neelance) 
> So... I went into the rabbit hole... It went very very deep... But I finally found the root cause!
> 
> npm-reqistry-fetch uses make-fetch-happen, see [here](https://github.com/npm/npm-registry-fetch/blob/e58e8bc6c5b25d53a764f58190fc3a5c764a2e78/lib/index.js#L6).
> make-fetch-happen uses minipass-fetch, see [here](https://github.com/npm/make-fetch-happen/blob/ed73ef528397a574df711e24238aaf7954512e55/lib/remote.js#L2).
> minipass-fetch uses Node's builtin http/https, see [here](https://github.com/npm/minipass-fetch/blob/f5f6e20b35d240b8f3ec220ef7d221b1488ce294/lib/index.js#L65).
> The maxSockets parameter gets passed through all these layers just fine, no issue here.
> However, the implementation of http.Agent in Node.js is suboptimal: It checks for maxSockets [here](https://github.com/nodejs/node/blob/fec7e505fc6f256522eb0031515757f2e58daa61/lib/_http_agent.js#L284), but does not immediately increase the checked value. Instead it calls this.createSocket and assumes the function pushes to this.sockets synchronously. Typical risk for a race condition.
> Now back up the stack again: make-fetch-happen uses @npmcli/agent, see [here](https://github.com/npm/make-fetch-happen/blob/ed73ef528397a574df711e24238aaf7954512e55/lib/remote.js#L7)
@npmcli/agent uses agent-base, see [here](https://github.com/npm/agent/blob/b95de7af38a2a79d54e16e9a0b5e1bc0b4c745b1/lib/agents.js#L10)
agent-base subclasses http.Agent and overwrites createSocket with an implementation that is asynchronous, see [here](https://github.com/TooTallNate/proxy-agents/blob/b5f94e3222c0aaa3bc56218ff125e2c56417c86e/packages/agent-base/src/index.ts#L80-L100).
>  🫠

**Comment by neelance** [comment]
I added a workaround to #7072.

**Comment by silverwind** [comment]
Can we close this duplicate issue in favor of https://github.com/npm/cli/issues/7072?

**Comment by ljharb** [comment]
Duplicate of #7072
